### PR TITLE
Refactor get topic names and types

### PR DESCRIPTION
**Pull request by wjwwood**
Connects to ros2/ros2#361

### CODE REVIEW

File: rmw/include/rmw/get_service_names_and_types.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ * \return `RMW_RET_INVALID_ARGUMENT` if any arguments are invalid, or
+ * \return `RMW_RET_BAD_ALLOC` if memory allocation fails, or
+ * \return `RMW_RET_ERROR` if an unspecified error occurs.
+ */
+RMW_PUBLIC
+RMW_WARN_UNUSED
+rmw_ret_t
+rmw_get_service_names_and_types(
+  const rmw_node_t * node,
+  rcutils_allocator_t * allocator,
```
Comment (Karsten1987) [inline rmw/include/rmw/get_service_names_and_types.h]: nitpick: could that be a `const` pointer?

File: rmw/include/rmw/names_and_types.h [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+ *
+ * \param[inout] names_and_types object to be finalized
+ * \returns `RMW_RET_OK` on successfully running the check, or
+ * \returns `RMW_RET_INVALID_ARGUMENT` if names_and_types is NULL, or
+ * \returns `RMW_RET_ERROR` when an unspecified error occurs.
+ */
+RMW_PUBLIC
+RMW_WARN_UNUSED
+rmw_ret_t
+rmw_names_and_types_fini(rmw_names_and_types_t * names_and_types);
```
Comment (Karsten1987) [inline rmw/include/rmw/names_and_types.h]: shouldn't that take an allocator as well?

File: rmw/src/names_and_types.c [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    if (rcutils_ret != RCUTILS_RET_OK) {
+      RMW_SET_ERROR_MSG(rcutils_get_error_string_safe())
+      return rmw_convert_rcutils_ret_to_rmw_ret(rcutils_ret);
+    }
+  }
+  if (names_and_types->types) {
+    // Use the allocator in the names string array
+    // (prevents this data structure from having to also store it)
+    names_and_types->names.allocator.deallocate(
+      names_and_types->types, names_and_types->names.allocator.state);
```
Comment (Karsten1987) [inline rmw/src/names_and_types.c]: can there be a mismatch between this allocator and the allocator used for the init function?

File: rmw/src/names_and_types.c [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    names_and_types->types = NULL;
+  }
+  // Cleanup names string array
+  rcutils_ret = rcutils_string_array_fini(&names_and_types->names);
```
Comment (Karsten1987) [inline rmw/src/names_and_types.c]: doesn't this call take an allocator as well?
https://github.com/ros2/rcutils/blob/master/include/rcutils/types/string_array.h#L86

File: rmw/src/names_and_types.c (hunk 3 above)
Comment (wjwwood) [inline rmw/src/names_and_types.c]: No, I don't think so.

File: rmw/include/rmw/get_service_names_and_types.h (hunk 1 above)
Comment (wjwwood) [inline rmw/include/rmw/get_service_names_and_types.h]: I'm not 100% sure. On the one hand the allocator's state is mutable (i.e. when someone calls `.allocate` it might change the internal state of the allocator), on the other the `const` in the strict functional sense (see: https://stackoverflow.com/questions/13181546/const-correctness-for-structs-with-pointers) will only restrict changing the function pointers in the allocator.

Here I was using non-const to indicate the allocator is not just "read" from, but instead it is "used" which in my mind is something like "written to".

In short, it could be but the rest of our code which takes an allocator pointer is already non-const to indicate the allocator might be mutated in the function.

If we can find reason is should be const, then I'd update all of them at once rather than just this function.
